R/addAlpha.R

```diff
@@ -574,56 +599,96 @@ setMethod("addAlpha", signature = c(x = "SummarizedExperiment"),
             "populated or with 'tree' provided separately.", call. = FALSE)
     }
     # Check for unsupported values (negative values)
-    if( any(assay(x, assay.type) < 0) ){
+    if( any(assay(x, assay.type) < 0 & !is.na(assay(x, assay.type))) ){
         ind <- detected[["non_neg"]]
         index_rm <- detected[!ind, "index"]
         detected <- detected[ind, ]
         if( length(index_rm) > 0 ){
             FUN <- if (nrow(detected) == 0) stop else warning
-            FUN("The following indices cannot be calculated due to unsupported 
-                values (negative values): ", paste(index_rm, collapse = ", "), 
+            FUN("The following indices cannot be calculated due to unsupported
+                values (negative values): ", paste(index_rm, collapse = ", "),
                 call. = FALSE)
         }
     }
+    # The result is a data.frame where each column represent single index to be
+    # estimated. Rows include the name of index and other information.
+    detected <- as.data.frame( t(detected) )
     return(detected)
 }
 
 # This function rarifies the data niter of times and calculates index for the
 # rarified data. The result is a mean of the iterations.
-#' @importFrom DelayedMatrixStats colMeans2
-.alpha_rarefaction <- function(
-        x, assay.type, niter, FUN, index, name, ...){
-    # Calculating the mean of the subsampled alpha estimates ans storing them
-    res <- lapply(seq(niter), function(i){
-        # Subsampling the counts from the original TreeSE object.
+#' @importFrom DelayedMatrixStats rowMeans2
+#' @importFrom BiocParallel bplapply
+.alpha_rarefaction <- function(x, assay.type, index, name, niter, BPPARAM, ...){
+    # Subsample the data and then calculate index based on the rarified data
+    res <- bplapply(seq(niter), function(i){
         x_sub <- rarefyAssay(x, assay.type = assay.type, verbose = FALSE, ...)
-        # Calculating the diversity indices on the subsampled object
-        x_sub <- do.call(FUN, args = list(
-            x_sub, assay.type = assay.type, index = index,
-            name = "rarefaction_temp_result", list(...)))
```

Comment:
This is causing the discrepancy in results. These old lines included 2 bugs. First of all, the alpha diversity index is calculated based on original table (assay.type), not subsampled ones (by default, it is called "subsampled"). Moreover, the parameters are not passed correctly with "...".